File: SuperClaude/Agents/registry.py
"""
Agent Registry for SuperClaude Framework.

Updated to work with the v7 tiered architecture:
- agents/core/     - 16 primary agents
- agents/traits/   - 8 composable modifiers
- agents/extensions/ - 7 domain specialists

This is a standalone implementation that doesn't depend on the archived SDK.
"""

from __future__ import annotations

import logging
import re
from pathlib import Path
from typing import Any, Optional

import yaml


# Find the SuperClaude root (contains CLAUDE.md)
def _find_superclaude_root() -> Path:
    """Find SuperClaude root by searching upward for CLAUDE.md."""
    current = Path(__file__).resolve()
    for parent in [current] + list(current.parents):
        if (parent / "CLAUDE.md").exists():
            return parent
    # Fallback
    return Path(__file__).parent.parent.parent


SUPERCLAUDE_ROOT = _find_superclaude_root()
AGENTS_DIR = SUPERCLAUDE_ROOT / "agents"


class AgentMarkdownParser:
    """Parser for agent markdown files with YAML frontmatter."""

    def parse_file(self, filepath: Path) -> Optional[dict[str, Any]]:
        """
        Parse a markdown file and extract frontmatter.

        Args:
            filepath: Path to the markdown file

        Returns:
            Dictionary of frontmatter data or None if parsing fails
        """
        try:
            content = filepath.read_text(encoding="utf-8")
            return self._extract_frontmatter(content)
        except Exception:
            return None

    def _extract_frontmatter(self, content: str) -> Optional[dict[str, Any]]:
        """Extract YAML frontmatter from markdown content."""
        if not content.startswith("---"):
            return None

        end_match = re.search(r"\n---\s*\n", content[3:])
        if not end_match:
            return None

        yaml_content = content[3:end_match.start() + 3]

        try:
            frontmatter = yaml.safe_load(yaml_content)
            if isinstance(frontmatter, dict):
                return frontmatter
            return None
        except yaml.YAMLError:
            return None


class AgentRegistry:
    """
    Agent registry for v7 tiered architecture.

    Scans the new directory structure:
    - agents/core/      - Core agents (priority 1)
    - agents/traits/    - Composable traits (not directly selectable)
    - agents/extensions/ - Domain extensions (priority 2)
    """

    def __init__(self, agents_dir: Path | None = None):
        """
        Initialize the agent registry with tiered agent paths.

        Args:
            agents_dir: Base agents directory. Defaults to SuperClaude/agents/.
        """
        if agents_dir is None:
            agents_dir = AGENTS_DIR

        self.agents_dir = agents_dir
        self.logger = logging.getLogger("agent.registry.v7")

        # Agent storage
        self._agents: dict[str, dict[str, Any]] = {}  # name -> config
        self._traits: dict[str, dict[str, Any]] = {}  # trait_name -> config
        self._categories: dict[str, list[str]] = {}  # category -> [agent_names]
        self._tier_agents: dict[str, list[str]] = {  # tier -> [agent_names]
            "core": [],
            "trait": [],
            "extension": [],
        }

        # Parser for markdown files
        self.parser = AgentMarkdownParser()

        # Discovery flag
        self._discovered = False

    def discover_agents(self, force: bool = False) -> int:
        """
        Discover all agents from the tiered directory structure.

        Args:
            force: Force re-discovery even if already discovered

        Returns:
            Number of agents discovered
        """
        if self._discovered and not force:
            return len(self._agents)

        self._agents.clear()
        self._categories.clear()
        self._traits.clear()
        self._tier_agents = {"core": [], "trait": [], "extension": []}

        total = 0

        # Discover core agents
        core_dir = self.agents_dir / "core"
        if core_dir.exists():
            count = self._discover_tiered_agents(core_dir, tier="core", is_core=True)
            total += count
            self.logger.debug(f"Discovered {count} core agents")

        # Discover traits (stored separately, not directly selectable)
        traits_dir = self.agents_dir / "traits"
        if traits_dir.exists():
            count = self._discover_tiered_agents(traits_dir, tier="trait", is_core=False)
            total += count
            self.logger.debug(f"Discovered {count} traits")

        # Discover extensions
        extensions_dir = self.agents_dir / "extensions"
        if extensions_dir.exists():
            count = self._discover_tiered_agents(extensions_dir, tier="extension", is_core=False)
            total += count
            self.logger.debug(f"Discovered {count} extension agents")

        self._discovered = True

        self.logger.info(
            f"Discovered {total} agents: "
            f"{len(self._tier_agents['core'])} core, "
            f"{len(self._tier_agents['trait'])} traits, "
            f"{len(self._tier_agents['extension'])} extensions"
        )

        return total

    def _discover_tiered_agents(
        self, directory: Path, tier: str, is_core: bool = False
    ) -> int:
        """
        Discover agents from a tiered directory.

        Args:
            directory: Directory to search
            tier: Agent tier (core, trait, extension)
            is_core: Whether these are core agents

        Returns:
            Number of agents discovered
        """
        count = 0

        for md_file in directory.glob("*.md"):
            try:
                config = self.parser.parse_file(md_file)
                if not config:
                    continue

                # Get agent name from frontmatter or filename
                name = config.get("name", md_file.stem)

                # Add tier information
                config["tier"] = tier
                config["is_core"] = is_core
                config["file_path"] = str(md_file)

                # Set priority based on tier
                if "priority" not in config:
                    if tier == "core":
                        config["priority"] = 1
                    elif tier == "extension":
                        config["priority"] = 2
                    else:  # trait
                        config["priority"] = 0  # Traits don't participate in selection

                # Store agent
                if tier == "trait":
                    self._traits[name] = config
                else:
                    self._agents[name] = config

                # Track by tier
                self._tier_agents[tier].append(name)

                # Track by category
                category = config.get("category", "general")
                if category not in self._categories:
                    self._categories[category] = []
                self._categories[category].append(name)

                count += 1

            except Exception as e:
                self.logger.warning(f"Failed to parse {md_file}: {e}")

        return count

    def get_all_agents(self) -> list[str]:
        """
        Get all selectable agent names (excludes traits).

        Returns:
            List of agent names
        """
        return list(self._agents.keys())

    def get_all_traits(self) -> list[str]:
        """
        Get all trait names.

        Returns:
            List of trait names
        """
        return list(self._traits.keys())

    def get_agent(self, name: str) -> Optional[dict[str, Any]]:
        """
        Get an agent by name.

        Args:
            name: Agent name

        Returns:
            Agent configuration or None
        """
        return self._agents.get(name)

    def get_agent_config(self, name: str) -> Optional[dict[str, Any]]:
        """
        Get configuration for a specific agent.

        Args:
            name: Agent name

        Returns:
            Agent configuration dictionary or None
        """
        return self._agents.get(name)

    def get_trait_config(self, trait_name: str) -> Optional[dict[str, Any]]:
        """
        Get configuration for a specific trait.

        Args:
            trait_name: Name of the trait

        Returns:
            Trait configuration dictionary or None
        """
        return self._traits.get(trait_name)

    def get_agents_by_tier(self, tier: str) -> list[str]:
        """
        Get agent names for a specific tier.

        Args:
            tier: Tier name (core, trait, extension)

        Returns:
            List of agent names in that tier
        """
        return self._tier_agents.get(tier, [])

    def get_agents_by_category(self, category: str) -> list[str]:
        """
        Get agent names for a specific category.

        Args:
            category: Category name

        Returns:
            List of agent names in that category
        """
        return self._categories.get(category, [])

    def is_valid_trait(self, trait_name: str) -> bool:
        """
        Check if a trait name is valid.

        Args:
            trait_name: Name to check

        Returns:
            True if trait exists
        """
        return trait_name in self._traits


__all__ = [
    "AgentRegistry",
    "AgentMarkdownParser",
    "SUPERCLAUDE_ROOT",
    "AGENTS_DIR",
]

File: crates/superclaude-runtime/src/registry.rs
//! Agent Registry for SuperClaude Framework.
//!
//! Updated to work with the v7 tiered architecture:
//! - agents/core/     - 16 primary agents
//! - agents/traits/   - 10 composable modifiers
//! - agents/extensions/ - 7 domain specialists

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use tracing::{debug, info, warn};
use walkdir::WalkDir;

/// Agent configuration extracted from frontmatter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentConfig {
    /// Agent name
    pub name: String,

    /// Description of agent purpose
    #[serde(default)]
    pub description: String,

    /// Agent tier (core, trait, extension)
    pub tier: String,

    /// Category (e.g., general, frontend, backend)
    #[serde(default)]
    pub category: String,

    /// Trigger keywords for agent selection
    #[serde(default)]
    pub triggers: Vec<String>,

    /// Available tools for this agent
    #[serde(default)]
    pub tools: Vec<String>,

    /// File patterns this agent handles
    #[serde(default)]
    pub file_patterns: Vec<String>,

    /// Selection priority (1=highest, lower=higher priority)
    #[serde(default = "default_priority")]
    pub priority: u32,

    /// Whether this is a core agent
    #[serde(default)]
    pub is_core: bool,

    /// Path to the agent markdown file
    #[serde(skip)]
    pub file_path: PathBuf,
}

fn default_priority() -> u32 {
    2
}

impl AgentConfig {
    /// Create a new AgentConfig from frontmatter and file path
    pub fn from_frontmatter(mut frontmatter: HashMap<String, serde_yaml::Value>, file_path: PathBuf) -> Result<Self> {
        // Extract name from frontmatter or use filename
        let name = frontmatter
            .get("name")
            .and_then(|v| v.as_str())
            .map(String::from)
            .unwrap_or_else(|| {
                file_path
                    .file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or("unknown")
                    .to_string()
            });

        // Get tier from frontmatter
        let tier = frontmatter
            .get("tier")
            .and_then(|v| v.as_str())
            .map(String::from)
            .unwrap_or_else(|| "core".to_string());

        // Determine is_core based on tier
        let is_core = tier == "core";

        // Set priority based on tier if not specified
        if !frontmatter.contains_key("priority") {
            let priority = match tier.as_str() {
                "core" => 1,
                "extension" => 2,
                _ => 0, // traits don't participate in selection
            };
            frontmatter.insert("priority".to_string(), serde_yaml::Value::Number(priority.into()));
        }

        // Add tier and is_core to frontmatter
        frontmatter.insert("tier".to_string(), serde_yaml::Value::String(tier));
        frontmatter.insert("is_core".to_string(), serde_yaml::Value::Bool(is_core));
        frontmatter.insert("name".to_string(), serde_yaml::Value::String(name));

        // Convert to AgentConfig
        let mut config: AgentConfig = serde_yaml::from_value(serde_yaml::Value::Mapping(
            frontmatter.into_iter().map(|(k, v)| {
                (serde_yaml::Value::String(k), v)
            }).collect()
        ))?;

        config.file_path = file_path;
        Ok(config)
    }
}

/// Agent registry for v7 tiered architecture
pub struct AgentRegistry {
    /// Base agents directory
    agents_dir: PathBuf,

    /// Registered agents (name -> config), excludes traits
    agents: HashMap<String, AgentConfig>,

    /// Registered traits (trait_name -> config)
    traits: HashMap<String, AgentConfig>,

    /// Agents by category
    categories: HashMap<String, Vec<String>>,

    /// Agents by tier
    tier_agents: HashMap<String, Vec<String>>,

    /// Whether discovery has been run
    discovered: bool,
}

impl AgentRegistry {
    /// Create a new agent registry
    pub fn new(agents_dir: Option<PathBuf>) -> Self {
        let agents_dir = agents_dir.unwrap_or_else(|| {
            // Find SuperClaude root by looking for CLAUDE.md
            Self::find_superclaude_root()
                .map(|root| root.join("agents"))
                .unwrap_or_else(|| PathBuf::from("agents"))
        });

        Self {
            agents_dir,
            agents: HashMap::new(),
            traits: HashMap::new(),
            categories: HashMap::new(),
            tier_agents: HashMap::from([
                ("core".to_string(), Vec::new()),
                ("trait".to_string(), Vec::new()),
                ("extension".to_string(), Vec::new()),
            ]),
            discovered: false,
        }
    }

    /// Find SuperClaude root by searching upward for CLAUDE.md
    fn find_superclaude_root() -> Option<PathBuf> {
        let mut current = std::env::current_dir().ok()?;

        loop {
            if current.join("CLAUDE.md").exists() {
                return Some(current);
            }

            if !current.pop() {
                break;
            }
        }

        None
    }

    /// Discover all agents from the tiered directory structure
    pub fn discover_agents(&mut self, force: bool) -> Result<usize> {
        if self.discovered && !force {
            return Ok(self.agents.len() + self.traits.len());
        }

        self.agents.clear();
        self.categories.clear();
        self.traits.clear();
        self.tier_agents = HashMap::from([
            ("core".to_string(), Vec::new()),
            ("trait".to_string(), Vec::new()),
            ("extension".to_string(), Vec::new()),
        ]);

        let mut total = 0;

        // Discover core agents
        let core_dir = self.agents_dir.join("core");
        if core_dir.exists() {
            let count = self.discover_tiered_agents(&core_dir, "core", true)?;
            total += count;
            debug!("Discovered {} core agents", count);
        }

        // Discover traits
        let traits_dir = self.agents_dir.join("traits");
        if traits_dir.exists() {
            let count = self.discover_tiered_agents(&traits_dir, "trait", false)?;
            total += count;
            debug!("Discovered {} traits", count);
        }

        // Discover extensions
        let extensions_dir = self.agents_dir.join("extensions");
        if extensions_dir.exists() {
            let count = self.discover_tiered_agents(&extensions_dir, "extension", false)?;
            total += count;
            debug!("Discovered {} extension agents", count);
        }

        self.discovered = true;

        info!(
            "Discovered {} agents: {} core, {} traits, {} extensions",
            total,
            self.tier_agents.get("core").map(|v| v.len()).unwrap_or(0),
            self.tier_agents.get("trait").map(|v| v.len()).unwrap_or(0),
            self.tier_agents.get("extension").map(|v| v.len()).unwrap_or(0)
        );

        Ok(total)
    }

    /// Discover agents from a tiered directory
    fn discover_tiered_agents(&mut self, directory: &Path, tier: &str, is_core: bool) -> Result<usize> {
        let mut count = 0;

        for entry in WalkDir::new(directory)
            .max_depth(1)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            let path = entry.path();

            // Only process .md files
            if path.extension().and_then(|s| s.to_str()) != Some("md") {
                continue;
            }

            match self.parse_agent_file(path, tier, is_core) {
                Ok(config) => {
                    let name = config.name.clone();

                    // Store agent or trait
                    if tier == "trait" {
                        self.traits.insert(name.clone(), config);
                    } else {
                        self.agents.insert(name.clone(), config.clone());

                        // Track by category
                        let category = config.category.clone();
                        self.categories
                            .entry(category)
                            .or_insert_with(Vec::new)
                            .push(name.clone());
                    }

                    // Track by tier
                    self.tier_agents
                        .entry(tier.to_string())
                        .or_insert_with(Vec::new)
                        .push(name);

                    count += 1;
                }
                Err(e) => {
                    warn!("Failed to parse agent file {:?}: {}", path, e);
                }
            }
        }

        Ok(count)
    }

    /// Parse an agent markdown file
    fn parse_agent_file(&self, path: &Path, _tier: &str, _is_core: bool) -> Result<AgentConfig> {
        let content = std::fs::read_to_string(path)
            .with_context(|| format!("Failed to read file: {:?}", path))?;

        // Extract frontmatter between --- markers (with multiline and dotall flags)
        let re = regex::Regex::new(r"(?ms)^---\s*\n(.*?)\n---\s*\n").unwrap();

        let frontmatter_str = re
            .captures(&content)
            .and_then(|caps| caps.get(1))
            .map(|m| m.as_str())
            .with_context(|| format!("No frontmatter found in {:?}", path))?;

        // Parse as YAML
        let yaml_value: serde_yaml::Value = serde_yaml::from_str(frontmatter_str)
            .with_context(|| format!("Failed to parse YAML frontmatter from {:?}", path))?;

        let frontmatter: HashMap<String, serde_yaml::Value> = match yaml_value {
            serde_yaml::Value::Mapping(map) => {
                map.into_iter()
                    .map(|(k, v)| {
                        let key = k.as_str().unwrap_or("").to_string();
                        (key, v)
                    })
                    .collect()
            }
            _ => anyhow::bail!("Frontmatter is not a mapping"),
        };

        AgentConfig::from_frontmatter(frontmatter, path.to_path_buf())
    }

    /// Get all selectable agent names (excludes traits)
    pub fn get_all_agents(&self) -> Vec<String> {
        self.agents.keys().cloned().collect()
    }

    /// Get all trait names
    pub fn get_all_traits(&self) -> Vec<String> {
        self.traits.keys().cloned().collect()
    }

    /// Get an agent by name
    pub fn get_agent(&self, name: &str) -> Option<&AgentConfig> {
        self.agents.get(name)
    }

    /// Get agent configuration by name
    pub fn get_agent_config(&self, name: &str) -> Option<&AgentConfig> {
        self.agents.get(name)
    }

    /// Get trait configuration by name
    pub fn get_trait_config(&self, trait_name: &str) -> Option<&AgentConfig> {
        self.traits.get(trait_name)
    }

    /// Get agent names for a specific tier
    pub fn get_agents_by_tier(&self, tier: &str) -> Vec<String> {
        self.tier_agents
            .get(tier)
            .cloned()
            .unwrap_or_default()
    }

    /// Get agent names for a specific category
    pub fn get_agents_by_category(&self, category: &str) -> Vec<String> {
        self.categories
            .get(category)
            .cloned()
            .unwrap_or_default()
    }

    /// Check if a trait name is valid
    pub fn is_valid_trait(&self, trait_name: &str) -> bool {
        self.traits.contains_key(trait_name)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    fn create_test_agent(dir: &Path, name: &str, tier: &str, triggers: &[&str]) -> Result<()> {
        let content = format!(
            r#"---
name: {}
description: Test agent
tier: {}
category: test
triggers: [{}]
---

# Test Agent
"#,
            name,
            tier,
            triggers.iter().map(|t| format!("\"{}\"", t)).collect::<Vec<_>>().join(", ")
        );

        fs::write(dir.join(format!("{}.md", name)), content)?;
        Ok(())
    }

    #[test]
    fn test_agent_discovery() -> Result<()> {
        let temp_dir = TempDir::new()?;
        let agents_dir = temp_dir.path();

        // Create directory structure
        fs::create_dir(agents_dir.join("core"))?;
        fs::create_dir(agents_dir.join("traits"))?;
        fs::create_dir(agents_dir.join("extensions"))?;

        // Create test agents
        create_test_agent(&agents_dir.join("core"), "test-core", "core", &["test", "core"])?;
        create_test_agent(&agents_dir.join("traits"), "test-trait", "trait", &[])?;
        create_test_agent(&agents_dir.join("extensions"), "test-ext", "extension", &["extension"])?;

        // Create registry and discover
        let mut registry = AgentRegistry::new(Some(agents_dir.to_path_buf()));
        let count = registry.discover_agents(false)?;

        assert_eq!(count, 3);
        assert_eq!(registry.get_all_agents().len(), 2); // core + extension
        assert_eq!(registry.get_all_traits().len(), 1);

        Ok(())
    }

    #[test]
    fn test_agent_retrieval() -> Result<()> {
        let temp_dir = TempDir::new()?;
        let agents_dir = temp_dir.path();

        fs::create_dir(agents_dir.join("core"))?;
        create_test_agent(&agents_dir.join("core"), "test-agent", "core", &["test"])?;

        let mut registry = AgentRegistry::new(Some(agents_dir.to_path_buf()));
        registry.discover_agents(false)?;

        let agent = registry.get_agent("test-agent");
        assert!(agent.is_some());

        let config = agent.unwrap();
        assert_eq!(config.name, "test-agent");
        assert_eq!(config.tier, "core");
        assert!(config.is_core);

        Ok(())
    }

    #[test]
    fn test_tier_filtering() -> Result<()> {
        let temp_dir = TempDir::new()?;
        let agents_dir = temp_dir.path();

        fs::create_dir(agents_dir.join("core"))?;
        fs::create_dir(agents_dir.join("extensions"))?;

        create_test_agent(&agents_dir.join("core"), "core-agent", "core", &[])?;
        create_test_agent(&agents_dir.join("extensions"), "ext-agent", "extension", &[])?;

        let mut registry = AgentRegistry::new(Some(agents_dir.to_path_buf()));
        registry.discover_agents(false)?;

        let core_agents = registry.get_agents_by_tier("core");
        assert_eq!(core_agents.len(), 1);
        assert!(core_agents.contains(&"core-agent".to_string()));

        let ext_agents = registry.get_agents_by_tier("extension");
        assert_eq!(ext_agents.len(), 1);
        assert!(ext_agents.contains(&"ext-agent".to_string()));

        Ok(())
    }
}
